### ISSUE TITLE
Direct Line 500 error retrieving messages

### ISSUE THREAD

**Issue by Stormridge** [Microsoft/BotBuilder#311]
Node versions:  botbuilder 0.10.2, botconnector 0.9.6, all other libraries are latest.  

I'm getting an HTTP 500 error when retrieving messages from:  /api/conversations/{conversationId}/messages

When I remove all Authentication headers from that GET, I do get the expected 403 error complaining that I don't have a valid token.  However, as soon as I add back in my valid token, I get the 500 error.

I've tried adding/removing various headers as well, but still get that 500 error.

My simple bot is doing the following sequence (successfully, I believe)
A. Getting new token from the REST interfaces.  HTTP 200, all good.

    tokenData is like "dI-fzw4Sa-Y.dAA.AQBHAFAAMQBxAGMAbQBIAG4AQQBaAA.RDZXRYWx0QE.lOWzjmfBCsk.OQJ21QviV..."

B. Using new token to start a conversation, as follows:

```
request({
        method: 'POST',
        url: base_url,
        headers: {
            'Ocp-Apim-Subscription-Key' : primaryAppSecret,
            'Content-Type': 'application/json; charset=utf-8',
            'Authorization': "BotConnector " + tokenData
        }
    }, function (error, response, body) {  // store conversation id here...... }
```
    200 success from that...    conversationId is like "AGP1qcmHnAZ"

C.  POST a new message to that conversation, using token and conversation id.

```
request({
        method: 'POST',
        url: base_url,
        headers: {
            'Ocp-Apim-Subscription-Key': primaryAppSecret,
            'Content-Type': 'application/json; charset=utf-8',
            Authorization: 'BotConnector ' + tokenData
        },
        body: JSON.stringify( 
            {
                conversationId: conversationID,
                from: userID,
                text: message,
            })
        }, function (error, response, body) {
```

    That's a happy 204 response.

D.  Check to see if there are messages available.  This results in HTTP 500. 
```
request({
        method: 'GET',
        url: base_url,
        headers: {
            Authorization: 'BotConnector ' + tokenData
        }
    }, function (error, response, body) { .........}
```

    URL is like https://directline.botframework.com/api/conversations/AGP1qcmHnAZ/messages

Appreciate any tips, thanks.

**Comment by DavePilks** [comment]
I am having similar issues using C#

**Comment by Stormridge** [comment]
Happy to attempt any workarounds for this -- currently blocked by it.

Have tried using the private key instead (no token) but get same results.

Thanks

**Comment by anthonyjesmok** [comment]
+1 for getting this fixed. Any workarounds as well that people have created would be appreciated.

**Comment by Stormridge** [comment]
1 month later and this API still returns error 500.  Any updates?

**Comment by dandriscoll** [comment]
You caught me right in the middle of our last deployment. :) This is working for me, although my client is a little bit different (still in Node, though). Can you take a look and see if the GET call in step D is working for you?

**Comment by brendanarnold** [comment]
I get the same error.

If I do step B (start conversation) followed by D (read messages) it works

If however I do B (start conversation) followed by C (post message) then D (read messages) it does not work.

I did some experiments and found that 'Ocp-Apim-Subscription-Key' is not necessary in POST and that a more fully formed body was causing the problems i.e. this worked ...

```
{
  "conversationId": "{{conversationId}}",
  "from": "someusername",
  "text": "annual leave",
}
```
 but this did not ...

```
{
  "id": "SomeId",
  "conversationId": "{{conversationId}}",
  "created": "{{date}}",
  "from": "someusername",
  "text": "annual leave",
  "channelData": "Conv1",
  "images": [],
  "attachments": [],
  "eTag": "blah"
}
```

where curly brackets substitute variables. 

I know you are already using the short form, but maybe check exactly what is being sent in your body? Maybe a substitution is not working? Maybe they are not being wrapped in double quotes?

**Comment by Stormridge** [comment]
Hi Dan, Can you share your sample code and sequence that makes this work correctly? 

My Step D (GET'ing messages) still does not work and throws a 500 error.

**Comment by dandriscoll** [comment]
I found an inconsistency in the configuration within the Bot Connector service that may have been causing this. I've fixed up all of the records.

Here's some sample TypeScript I use to POST a message to a conversation:

```
    import https = require('https');
    function postMessage(conversationId: string, secretOrToken: string, message: any, callback: (err: any, res: any) => void) {
        var body = JSON.stringify(message);
        var request = https.request({
            method: 'POST',
            host: 'directline.botframework.com',
            port: 443,
            path: '/api/conversations/' + conversationId + '/messages',
            headers: {
                'Authorization': 'BotConnector ' + secretOrToken,
                'Content-Length': body.length,
                'Content-Type': 'application/json; charset=utf-8'
            }
        }, res => {
            var data = '';
            res.on('data', chunk => {
                data += chunk;
            }).on('error', err => {
                callback(err, null);
            }).on('end', () => {
                callback(null, data ? JSON.parse(data) : null);
            });
        });
        request.write(body);
        request.end();
    }
    postMessage(CONVERSATIONID, SECRET, { 'text': 'hello', 'from': 'user0' }, CALLBACK);
```

**Comment by Stormridge** [comment]
Thanks Dan.

The problem still is, everything works great until we need to GET messages, which results in HTTP 500.  When we post a message to a conversation, the server is happy and returns 204.

But a simple GET of that conversation's messages always returns error 500.

base_url is like https://directline.botframework.com/api/conversations/AGP1qcmHnAZ/messages

    request({
        method: 'GET',
        url: base_url,
        port: 443,
        headers: {
            'Authorization': 'BotConnector ' + directLineSecret,
            'Content-Type': 'application/json; charset=utf-8'
        }
    }, function (error, response, body) {
           ...

That simple snippet returns 500, even after a successful 203 posting of a message to that conversation.

**Comment by v-thgratMSGit** [comment]
Closing this issue due to age. @Stormridge - please open a new issue if this issue is not fixed and needs further discussion. Thanks!

**Comment by yonahforst** [comment]
For anyone stumbling across this, I was also getting 500 errors posting to the `/v3/conversations` endpoint. I solved it by making sure the `Content-Length` headers were set

**Comment by chris31389** [comment]
My issue was that I didn't send the request with content-type: "application/json" hope this helps!
